### PR TITLE
Overview function with normalised axis (for cases and deaths)

### PR DESCRIPTION
**Pull request by kirienko**
Here is the attempt of having two Y-axes on the same plot: one is "normal" and another is normalized per 100K people.

There is one thing we didn't think about in advance: in our 14-day incidence rate we _accumulate_ cases, although here we just normalise daily changes. Here what we actually do is we change the units, but the graph preserve its shape. If want to have integral metric (like 14-days), it will be a completely different graph.

The question is: is that type of normalization I introduce in this commit is what we want? If yes, then I can think about more intuitive representations of the two grids.

image

### CODE REVIEW

**Comment by fangohr** [comment]
Hi Yury, yes, I was thinking of not changing the data (i.e. no integration / sum over window) and the shape will stay the same.

**Comment by kirienko** [comment]
Sorry for the delay with the solution. The reason was mainly our famous PR with the 14-days incidence rate. There we [had broken](https://github.com/oscovida/oscovida/pull/102/files#diff-0a2ce53910eb6d0d4be75a50ccfc3989) some important stuff, namely the `get_population` function. Initially it supported countries and regions within countries, but then it was removed for some good reasons. And since some parts were not covered with tests, we didn't noticed that. (My bad, I personally gave an LGTM for that.)

Here I fixed some of those issues (and probably introduces some new ones :smile:):
* I restored initial `get_population` function under the name `population`. Unfortunately it only supports for Germany (i.e. _Bundesländer_) but not, say, states within the US;
* I fixed broken `normalise` feature for `make_compare_plot`;
* [by-product] I also tried to decouple the fetching data and the label creation (see `get_region_label`) — I fully agree that it'd be better to do a separate PR for that, but I decided that I've already delayed for too long, and that's why I decided not to spend additional time for separating that piece of code;
* the tests were updated accordingly.

**Comment by kirienko** [comment]
Damn, the test for `make_compare_plot_hungary()` is failing now. Have no idea why, but it's not me!

**Comment by fangohr** [comment]
Proposed fine tuning of plots (from discussion):

- put relative incidence on left y-axis, leave matplotlib to choose sensible labels
- put total numbers on the right y-axis
- use the gridlines from left axis for right axis, by using non-round numbers on the right

**Comment by fangohr** [comment]
- double check plot on title page (https://oscovida.github.io/staging/111/index.html): overlap of y-axis label on the right with numbers

File: oscovida/oscovida.py [hunk 1]
```diff
@@ -478,6 +478,39 @@ def get_population() -> pd.DataFrame:
...
+            else:
+                raise NotImplementedError(f"{region} in neither in available German Lands nor in Landkreises. " \
+                                          f"These are {', '.join(sorted(df['BL'].drop_duplicates()))} for Lands and " \
+                                          f"{', '.join(sorted(df.index))} for Landkreises.")
+    else:
+        df = get_population()
+        if country in df.index and region is None:
+            return int(df.loc[country].population)
+        else:
+            return None
```
Comment (ebadkamil) [inline oscovida/oscovida.py:512]: `return`

File: oscovida/oscovida.py [hunk 2]
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
     """
     bar_alpha = 0.2
     if labels is None:
-        label = ""
-        region = ""
+        type_label = ""
```
Comment (ebadkamil) [inline oscovida/oscovida.py:786]:
```python
type_label = ""
region_label = ""
if labels:
    region_label, type_label = labels



```

File: oscovida/oscovida.py [hunk 3]
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
...
+        region_label = ""
     else:
-        region, label = labels
+        region_label, type_label = labels
+    ax_label = region_label + " new " + type_label
```
Comment (ebadkamil) [inline oscovida/oscovida.py:790]: `ax_label = f"{region_label} new {type_label}"`

File: oscovida/oscovida.py [hunk 4]
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
...
-    ax.bar(change.index, change.values, color=color,
-           label=ax_label, alpha=bar_alpha, linewidth=LW)
-
-    ax.plot(smooth2.index, smooth2.values, color=color,
-            label=ax_label + " " + smooth2_label, linewidth=LW)
+    if country and region:
+        habitants = population(country=country, region=region)
+    else:
+        habitants = population(region_label)
+    if region_label is not "" and habitants:
```
Comment (ebadkamil) [inline oscovida/oscovida.py:798]: `if region_label and habitants:`

File: oscovida/oscovida.py [hunk 5]
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
...
 
+def get_region_label(country: str, region: str = None, subregion: str = None) -> str:
+    """
+    Produce unique and unambiguous name from region and subregion
+    """
+    # TODO: this might be used to decouple the label creation and fetching data
+    _country = country.casefold()
+    region_label = None
+    if _country == 'germany':
+        # if region == None and subregion == None:
```
Comment (ebadkamil) [inline oscovida/oscovida.py:1175]: remove commented code

File: oscovida/oscovida.py [hunk 6]
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
...
+        #     region_label = _country
+        if region is not None and subregion is None:
+            region_label = f"Germany-{region}"
+        elif region is None and subregion is not None:
+            region_label = f"Germany-{subregion}"
+        elif region is None and subregion is not None:
```
Comment (ebadkamil) [inline oscovida/oscovida.py:1181]: repeated as above

File: oscovida/oscovida.py [hunk 7]
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
...
+            region_label = f"Germany-{subregion}"
+
+    elif _country.lower() == 'us' and region != None:
```
Comment (ebadkamil) [inline oscovida/oscovida.py:1184]: `elif _country.lower() == 'us' and region is not None:`

File: oscovida/oscovida.py [hunk 8]
```diff
@@ -1577,13 +1652,24 @@ def has_twin(ax: plt.Axes) -> bool:
     return False
 
 
+def align_twinx_ticks(ax_left: plt.Axes, ax_right: plt.Axes) -> np.ndarray:
+    # there's no easy way of aligning ticks nor a good general solution, therefore let's do a trick
+    # see here: https://stackoverflow.com/questions/45037386/trouble-aligning-ticks-for-matplotlib-twinx-axes
+    # obviously, f(left_min) -> right_min and f(left_max) -> right_max, and points in between are mapped proportionally
+    left = ax_left.get_ylim()
+    right = ax_right.get_ylim()
+    f = lambda x: right[0] + (x - left[0]) / (left[1] - left[0]) * (right[1] - right[0])
```
Comment (ebadkamil) [inline oscovida/oscovida.py:1661]: `return right[0] + (ax_left.get_yticks() - left[0]) / (left[1] - left[0]) * (right[1] - right[0])`

File: tests/test_corona.py [hunk 9]
```diff
@@ -81,12 +81,11 @@ def test_fetch_data_hungary():
 def test_choose_random_counties():
     # Hungary related
     with_local = c.choose_random_counties(exclude_region="Baranya", size=18)
-    print(with_local)
+    # print(with_local)
```
Comment (ebadkamil) [inline tests/test_corona.py:84]: remove commented code

**Review by ebadkamil — COMMENTED** [review]
besides minor comments LGTM.

**Comment by ebadkamil** [comment]
also commits can be squashed.

File: oscovida/oscovida.py [hunk 10]
```diff
@@ -736,7 +769,8 @@ def compute_daily_change(series):
     return change, smooth, smooth2
 
 
-def plot_daily_change(ax, series, color, labels=None):
+def plot_daily_change(ax, series: pd.Series, color: str, labels: Tuple[str, str] = None,
```
Comment (ebadkamil) [inline oscovida/oscovida.py:772]:
```python
def plot_daily_change(ax: plt.Axes, series: pd.Series, color: str, labels: Tuple[str, str] = None,
                                    country: str = None, region: str = None, subregion: str = None):
```

**Pull request merged**